flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoderDelegateTest.java

```diff
@@ -0,0 +1,411 @@
+				sizes[0] + sizes[1] + sizes[2],
+				sizes[0] + sizes[1] + sizes[2] + sizes[3],
+				sizes[0] + sizes[1] + sizes[2] + sizes[3] + sizes[4] / 3 * 2,
+				sizes[0] + sizes[1] + sizes[2] + sizes[3] + sizes[4] + sizes[5] + sizes[6] / 2
+			});
+
+		repartitionMessagesAndVerifyDecoding(
+			context,
+			messagesSupplier,
+			(int[] sizes) -> new int[]{
+				sizes[0] / 3,
+				sizes[0],
+				sizes[0] + sizes[1] / 3,
+				sizes[0] + sizes[1] / 3 * 2,
+				sizes[0] + sizes[1] + sizes[2] / 3,
+				sizes[0] + sizes[1] + sizes[2] + sizes[3] + sizes[4] / 3,
+				sizes[0] + sizes[1] + sizes[2] + sizes[3] + sizes[4] + sizes[5] + sizes[6]
+			});
+	}
+
+	//------------------------------------------------------------------------------------------------------------------
+
+	private void repartitionMessagesAndVerifyDecoding(
+		NettyChannelAndInputChannelIds context,
+		Supplier<NettyMessage[]> messagesSupplier,
+		Function<int[], int[]> rePartitioner) throws Exception {
+
+		NettyMessage[] messages = messagesSupplier.get();
+
+		ByteBuf[] serializedBuffers = null;
+		ByteBuf mergedBuffer = null;
+
+		try {
+			serializedBuffers = serializeMessages(messages);
+			int[] sizes = getBufferSizes(serializedBuffers);
+			mergedBuffer = mergeBuffers(serializedBuffers);
+
+			ByteBuf[] partitionedBuffers = partitionBuffer(mergedBuffer, rePartitioner.apply(sizes));
+			List<NettyMessage> decodedMessages = decodedMessages(context, partitionedBuffers);
+
+			verifyDecodedMessages(context, messages, decodedMessages);
+		} finally {
+			if (serializedBuffers != null) {
+				releaseBuffers(serializedBuffers);
+			}
+
+			if (mergedBuffer != null) {
+				mergedBuffer.release();
+			}
+		}
+	}
+
+	private ByteBuf[] serializeMessages(NettyMessage[] messages) throws Exception {
+		ByteBuf[] serializedBuffers = new ByteBuf[messages.length];
+		for (int i = 0; i < messages.length; ++i) {
+			serializedBuffers[i] = messages[i].write(ALLOCATOR);
+		}
+
+		return serializedBuffers;
+	}
+
+	private int[] getBufferSizes(ByteBuf[] buffers) {
+		int[] sizes = new int[buffers.length];
+		for (int i = 0; i < sizes.length; ++i) {
+			sizes[i] = buffers[i].readableBytes();
+		}
+
+		return sizes;
+	}
+
+	private ByteBuf mergeBuffers(ByteBuf[] buffers) {
+		ByteBuf allData = ALLOCATOR.buffer();
+		for (ByteBuf buffer : buffers) {
+			allData.writeBytes(buffer);
+		}
+
+		return allData;
+	}
+
+	private ByteBuf[] partitionBuffer(ByteBuf buffer, int[] partitionPositions) {
+		ByteBuf[] result = new ByteBuf[partitionPositions.length + 1];
+		for (int i = 0; i < result.length; ++i) {
+			int startPos = (i == 0 ? 0 : partitionPositions[i - 1]);
+			int endPos = (i < partitionPositions.length ? partitionPositions[i] : buffer.readableBytes());
+
+			result[i] = ALLOCATOR.buffer();
+			result[i].writeBytes(buffer, startPos, endPos - startPos);
+		}
+
+		return result;
+	}
+
+	private List<NettyMessage> decodedMessages(NettyChannelAndInputChannelIds context, ByteBuf[] inputBuffers) {
```

Comment:
NettyChannelAndInputChannelIds -> EmbeddedChannel
I do not suggesting putting `EmbeddedChannel` inside `NettyChannelAndInputChannelIds`, and it is a bit over-design.

Reply:
Removed `NettyChannelAndInputChannelIds` and use the fixed input channel ids.